pandas/core/internals/array_manager.py

```diff
@@ -339,9 +341,24 @@ def where(self, other, cond, align: bool, errors: str, axis: int) -> ArrayManage
             axis=axis,
         )
 
-    # TODO what is this used for?
-    # def setitem(self, indexer, value) -> ArrayManager:
-    #     return self.apply_with_block("setitem", indexer=indexer, value=value)
+    def setitem(self, indexer, value, column_idx) -> ArrayManager:
+        """
+        Set value for a single column and a given row indexer. For example, from
+        ``df.loc[indexer] = value``
+        """
+        arr = self.arrays[column_idx]
+
+        # TODO this special case can be removed once we only store EAs
+        # special case to support setting np.nan in a non-float numpy array
+        if (
+            isinstance(arr, np.ndarray)
+            and is_datetime64_dtype(arr.dtype)
+            and is_scalar(value)
+            and isna(value)
```

Comment:
pls use is_valid_nat_for_dtype, otherwise this will incorrectly let through td64 nat 

(i know i know, this is just  draft and youre not looking for comments like this, but this could easily fall through the cracks)

Reply:
No, this is certainly a comment that is useful ;) (I was actually planning to comment on this and ask you, because I already realized the same would need to be done for timedelta, that just didn't occur in the tests)

Reply:
Actually, in this case we explicitly want to include np.nan (which I think `is_valid_nat_for_dtype` doesn't do?), since we allow setting np.nan or None in a datetime column, but a datetime64[ns] numpy array doesn't allow this (and the same for timedelta64)

Reply:
> which I think is_valid_nat_for_dtype doesn't do?

incorrect